solr/core/src/java/org/apache/solr/cloud/api/collections/DeleteCollectionCmd.java

```diff
@@ -149,7 +150,15 @@ public void call(ClusterState state, ZkNodeProps message, @SuppressWarnings({"ra
       ocmh.overseer.offerStateUpdate(Utils.toJSON(m));
 
       // wait for a while until we don't see the collection
-      zkStateReader.waitForState(collection, 60, TimeUnit.SECONDS, (collectionState) -> collectionState == null);
+      try {
+        zkStateReader.waitForState(collection, 60, TimeUnit.SECONDS, (collectionState) -> collectionState == null);
+      } catch (SolrException e) {
+        if (e.getCause() instanceof PerReplicaStates.PrsZkNodeNotFoundException) {
+          log.info("PRS entry for collection " + collection + " not found in ZK. It was probably deleted between state.json read and PRS entry read.");
+        } else {
```

Comment:
Yeah. something like this. `DeleteCollectionCmd` is just waiting for collection to remove from ZK. look these lines in stacktrace. See if this can be change to verify that (collection has been removed)?
```
at org.apache.solr.common.cloud.ZkStateReader.waitForState(ZkStateReader.java:2096)
	at org.apache.solr.cloud.api.collections.DeleteCollectionCmd.call(DeleteCollectionCmd.java:152)
```

Reply:
Wondering if we have another `waitForState` method, which looks zk collection node(zk/solr/collection). And checks if collection has been removed. I think that would be clean solution.

Reply:
We probably could! We will likely need to re-implement the "watch" mechanism (since there's only DocCollectionWatcher and CollectionStateWatcher which both watch `DocCollection`), or use some sort of sleep loop?), which could be a bit complicated. It might not be entirely obvious if we provide 2 different overload versions on `waitForState` while one of them might break for PRS operations (the original signature one)

The new proposed fix is a bit lower level which does have the advantage to prevent higher level logic needing some "special knowledge/logic" to handle PRS race condition (which my previous DeleteCollectionCmd fix did 👎🏼 )

Reply:
I was thinking about `CloudCollectionsListener`. There might be some helper for this

Reply:
Ah for sure `CloudCollectionsListener` appears to provide notification on changes, and we might be able to implement another `waitForState` with similar "latch/wait" approach (or even refactor the existing signature to accept a more generic predicate).

I do still have concerns of having 2 different version of `waitForState` and the fact that higher level logic need to know the PRS race condition details to pick the "right" version to use. While the fix in `ZkStateReader` provides a more "complete" fix that higher level logic shielded from the race condition details, it should fix all the code reference to `ZkStateReader.fetchCollectionState`.

But of course the risk is that if the fix is incorrect (or trigger other issues), then more code would be affected 😓

Reply:
>>I do still have concerns of having 2 different version of waitForState and the fact that higher level logic need to know the PRS race condition details to pick the "right" version to use. While the fix in ZkStateReader provides a more "complete" fix that higher level logic is shielded from such lower level details and it should fix all the code reference to ZkStateReader.fetchCollectionState.

What exactly code is trying to do ? is it just waiting for collection to be removed from ZK - then it should be collection-type agnostic?

Reply:
yes it was waiting for the corresponding state.json entry be removed from ZK but ran into race condition as `ZkStateReader#waitForState` eventually calls `fetchCollectionState` that could have such race condition.

The current fix no longer targets the `DeleteCmd` as the first proposed fix, instead it now targets the `ZkStateReader#fetchCollectionState` directly and ensures race condition from state.json/PRS deletion would not cause failures.

Therefore, it should correctly detect the deletion in any cases such that `ZkStateReader#waitForState` should work properly.

Reply:
Right, if we have waitForState variant for `CloudCollectionsListener`, then implementation of `deletdCmd` will NOT depend on prs/nonprs collection. This listener will provide whether collection has been removed from ZK or not. Just wanted to make sure whether this make sense or not as haven't look all details?